src/dfx/src/commands/canister/snapshot.rs

```diff
@@ -206,3 +249,522 @@ async fn list(env: &dyn Environment, canister: String, call_sender: &CallSender)
     }
     Ok(())
 }
+
+async fn download(
+    env: &dyn Environment,
+    canister: String,
+    snapshot: SnapshotId,
+    dir: PathBuf,
+    call_sender: &CallSender,
+) -> DfxResult {
+    check_dir(&dir)?;
+
+    let canister_id = canister
+        .parse()
+        .or_else(|_| env.get_canister_id_store()?.get(&canister))?;
+
+    let retry_policy = ExponentialBackoff::default();
+
+    // Store metadata.
+    let metadata = retry(retry_policy.clone(), || async {
+        match read_canister_snapshot_metadata(env, canister_id, &snapshot.0, call_sender).await {
+            Ok(metadata) => Ok(metadata),
+            Err(_error) => Err(backoff::Error::transient(anyhow!(
+                "Failed to read metadata from snapshot {snapshot} in canister {canister}"
+            ))),
+        }
+    })
+    .await?;
+    let metadata_file = dir.join("metadata.json");
+    let metadata_json = serde_json::to_string_pretty(&metadata)?;
+    std::fs::write(&metadata_file, metadata_json).with_context(|| {
+        format!(
+            "Failed to write snapshot metadata to '{}'",
+            metadata_file.display()
+        )
+    })?;
+    debug!(
+        env.get_logger(),
+        "Snapshot metadata saved to '{}'",
+        metadata_file.display()
+    );
+
+    // Store Wasm module.
+    let wasm_module = read_blob(
+        env,
+        &canister,
+        canister_id,
+        &snapshot,
+        BlobKind::WasmModule,
+        metadata.wasm_module_size as usize,
+        retry_policy.clone(),
+        call_sender,
+    )
+    .await
+    .with_context(|| {
+        format!(
+            "Failed to read Wasm module from snapshot {snapshot} in canister {}",
+            canister_id.to_text(),
+        )
+    })?;
+    let wasm_module_file = dir.join("wasm_module.bin");
+    std::fs::write(&wasm_module_file, &wasm_module).with_context(|| {
+        format!(
+            "Failed to write Wasm module to '{}'",
+            wasm_module_file.display()
+        )
+    })?;
+    debug!(
+        env.get_logger(),
+        "Wasm module saved to '{}'",
+        wasm_module_file.display()
+    );
+
+    // Store Wasm memory.
+    let wasm_memory = read_blob(
+        env,
+        &canister,
+        canister_id,
+        &snapshot,
+        BlobKind::MainMemory,
+        metadata.wasm_memory_size as usize,
+        retry_policy.clone(),
+        call_sender,
+    )
+    .await
+    .with_context(|| {
+        format!(
+            "Failed to read Wasm memory from snapshot {snapshot} in canister {}",
+            canister_id.to_text(),
+        )
+    })?;
+    let wasm_memory_file = dir.join("wasm_memory.bin");
+    std::fs::write(&wasm_memory_file, &wasm_memory).with_context(|| {
+        format!(
+            "Failed to write Wasm memory to '{}'",
+            wasm_memory_file.display()
+        )
+    })?;
+    debug!(
+        env.get_logger(),
+        "Wasm memory saved to '{}'",
+        wasm_memory_file.display()
+    );
```

Comment:
I'd recommend refactoring this logic for dealing with binary data (wasm module, wasm memory, and stable memory) into a single reusable function.